### PR TITLE
royalties implemented

### PR DESCRIPTION
**Pull request by suNrisEinMyeYes**
-reworked payment processing
-implemented royalties via erc2981 standard
-tests for royalties done

### CODE REVIEW

File: contracts/marketplace/CruzoMarket.sol [hunk 1]
```diff
@@ -407,4 +405,23 @@ contract CruzoMarket is
...
+
+    function _paymentProcessing(
+        address _tokenAddress,
+        address _seller,
+        uint256 _tokenId,
+        uint256 _value
+    ) internal {
+        (address royaltyReciever, uint256 royaltyAmount) = IERC2981Upgradeable(
+            _tokenAddress
+        ).royaltyInfo(_tokenId, _value * (10000 - uint256(serviceFee)) / 10000);
```
Comment (bobrom) [inline contracts/marketplace/CruzoMarket.sol]: We may save the result of `_value * (10000 - uint256(serviceFee)) / 10000` in a variable, to not calculate it twice.

File: contracts/tokens/Cruzo1155.sol [hunk 2]
```diff
@@ -147,4 +149,36 @@ contract Cruzo1155 is Initializable, ERC1155URI {
...
+        onlyOwner
+    {
+        require(
+            _royaltyFeesInBips <= 5000,
+            "Royalty value must be between 0% and 50%"
+        );
+        _setDefaultRoyalty(_receiver, _royaltyFeesInBips);
+    }
+
+    function setTokenRoyalty(
```
Comment (bobrom) [inline contracts/tokens/Cruzo1155.sol:165]: I think we need to invoke `setTokenRoyalty` also within `create` function. This way we don't have to run separate transaction after token creation.

File: contracts/tokens/Cruzo1155.sol [hunk 3]
```diff
@@ -147,4 +149,36 @@ contract Cruzo1155 is Initializable, ERC1155URI {
...
+        address _receiver,
+        uint96 _royaltyFeesInBips,
+        uint256 _tokenId
+    ) public onlyOwner {
```
Comment (bobrom) [inline contracts/tokens/Cruzo1155.sol]: I'm not sure about `onlyOwner`. It'll work for collections which are created and owned by users. This way users will not have a possibility to control royalties when creating tokens in the default Cruzo collection, which is owned by us. Should not we use `onlyCreator` instead?

File: contracts/tokens/Cruzo1155.sol [hunk 4]
```diff
@@ -147,4 +149,36 @@ contract Cruzo1155 is Initializable, ERC1155URI {
...
+        require(
+            _royaltyFeesInBips <= 5000,
+            "Royalty value must be between 0% and 50%"
+        );
+
```
Comment (bobrom) [inline contracts/tokens/Cruzo1155.sol:174]: Is it normal that we allow users to adjust royalties anytime during the token lifecycle? Is it allowed on other NFT marketplaces?

**Review by bobrom — CHANGES_REQUESTED** [review]

File: contracts/tokens/Cruzo1155.sol (hunk 4 above)
Comment (suNrisEinMyeYes) [reply contracts/tokens/Cruzo1155.sol:174]: I will research this moment

Comment (suNrisEinMyeYes) [reply contracts/tokens/Cruzo1155.sol:174]: I can propose a solution: set fee only by creating process

**Review by kewka — APPROVED** [review]

File: contracts/tokens/Cruzo1155.sol [hunk 5]
```diff
@@ -147,4 +153,36 @@ contract Cruzo1155 is Initializable, ERC1155URI {
...
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        override(ERC2981Upgradeable, ERC1155Upgradeable)
+        returns (bool)
+    {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function setDefaultRoyaltyInfo(address _receiver, uint96 _royaltyFeesInBips)
```
Comment (bobrom) [inline contracts/tokens/Cruzo1155.sol]: @suNrisEinMyeYes how/when this "default" royalty is actually used in our case?

File: contracts/tokens/Cruzo1155.sol [hunk 6]
```diff
@@ -147,4 +153,36 @@ contract Cruzo1155 is Initializable, ERC1155URI {
...
+        public
+        onlyOwner
+    {
+        require(
```
Comment (bobrom) [inline contracts/tokens/Cruzo1155.sol]: @suNrisEinMyeYes Would it make sense to introduce a modifier with this `require`?

Comment (bobrom) [reply contracts/tokens/Cruzo1155.sol]: Can be done later with low priority.

**Review by bobrom — CHANGES_REQUESTED** [review]
We decided to not use default royalties. We only allow to set royalty when creating tokenId.

**Review by bobrom — APPROVED** [review]

**Pull request merged**